File: day3/solution part 2.cpp
#include <iostream>
#include <regex>
#include <string>
#include <fstream>

int main(){

    std::ifstream file("data.txt");
    std::string content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());

    file.close();

    std::regex pattern(R"(mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\))");

    std::sregex_iterator begin(content.begin(), content.end(), pattern);
    std::sregex_iterator end;

    long long int sum{};
    bool enable_multiplication = true;

    for (auto it = begin; it != end; ++it) {
        std::smatch match = *it;

        if (match.str(0) == "do()") {
            enable_multiplication = true;
        } 
        
        else if (match.str(0) == "don't()") {
            enable_multiplication = false;
        } 
        
        else if (enable_multiplication) {
            sum += std::stoi(match.str(1)) * std::stoi(match.str(2));
        }
    }
    std::cout << sum;
    return 0;
}

File: day3/solution part 2.py
import re

with open('data.txt', 'r') as file:
    content = file.read()

# pattern = r"mul\((\d+),(\d+)\)"
pattern = r"(mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\))"

enable_multiplication = True
 
sum = 0

data = re.findall(pattern, content)
for pair in data:
    match pair[0]:
        case "do()":
            enable_multiplication = True
        case "don't()":
            enable_multiplication = False
        case _ if enable_multiplication:
            sum += int(pair[1]) * int(pair[2])

print(sum)